packages/unified-selection/src/unified-selection/iModel/IModelSelectionHandler.ts

```diff
@@ -0,0 +1,221 @@
+
+  private _isSuspended: boolean;
+  private _cancelOngoingChanges = new Subject<void>();
+  private _unifiedSelectionListenerDisposeFunc: () => void;
+  private _iModelListenerDisposeFunc: () => void;
+
+  public constructor(props: IModelSelectionHandlerProps) {
+    this._iModel = props.iModel;
+    this._selectionStorage = props.selectionStorage;
+    this._cachingHiliteSetProvider = props.cachingHiliteSetProvider;
+    this._activeScopeProvider = props.activeScopeProvider;
+    this._queryExecutor = props.queryExecutor;
+    this._isSuspended = false;
+
+    this._hiliteSetProvider = createHiliteSetProvider({ queryExecutor: props.queryExecutor, metadataProvider: props.metadataProvider });
+    this._iModelListenerDisposeFunc = this._iModel.selectionSet.onChanged.addListener(this.onIModelSelectionChanged);
+    this._unifiedSelectionListenerDisposeFunc = this._selectionStorage.selectionChangeEvent.addListener(this.onUnifiedSelectionChanged);
+
+    // stop imodel from syncing tool selection with hilited list - we want to override that behavior
+    this._iModel.hilited.wantSyncWithSelectionSet = false;
+    this.applyCurrentHiliteSet();
+  }
+
+  public dispose() {
+    this._cancelOngoingChanges.next();
+    this._iModelListenerDisposeFunc();
+    this._unifiedSelectionListenerDisposeFunc();
+  }
+
+  /** Temporarily suspends tool selection synchronization until the returned `IDisposable` is disposed. */
+  public suspendIModelToolSelectionSync() {
+    const wasSuspended = this._isSuspended;
+    this._isSuspended = true;
+    return {
+      dispose: () => (this._isSuspended = wasSuspended),
+    };
+  }
+
+  private handleUnifiedSelectionChange(changeType: StorageSelectionChangeType, selectables: Selectables) {
+    if (changeType === "clear" || changeType === "replace") {
+      this.applyCurrentHiliteSet();
+      return;
+    }
+
+    if (changeType === "add") {
+      from(this._hiliteSetProvider.getHiliteSet({ selectables }))
+        .pipe(takeUntil(this._cancelOngoingChanges))
+        .subscribe({
+          next: (set) => {
+            this.applyHiliteSet(set);
+          },
+        });
+      return;
+    }
+
+    from(this._hiliteSetProvider.getHiliteSet({ selectables }))
+      .pipe(takeUntil(this._cancelOngoingChanges))
+      .subscribe({
+        next: (set) => {
+          this.removeHiliteSet(set);
+        },
+        complete: () => {
+          this.applyCurrentHiliteSet(false);
+        },
+      });
+  }
+
+  private onUnifiedSelectionChanged = (args: StorageSelectionChangeEventArgs) => {
+    // iModels are only interested in top-level selection changes
+    if (args.iModelKey !== this._iModel.key || args.level !== 0) {
+      return;
+    }
+
+    this._cancelOngoingChanges.next();
+    this.handleUnifiedSelectionChange(args.changeType, args.selectables);
+  };
+
+  private applyCurrentHiliteSet(clearBefore = true) {
+    if (clearBefore) {
```

Comment:
It's unclear what the argument means in the call sites - the following should help:

```suggestion
  private applyCurrentHiliteSet({ activeSelectionAction }: { activeSelectionAction: "clear" | "keep" }) {
    if (activeSelectionAction === "clear") {
```

Reply:
Changed.